### PR TITLE
fix: `Parse.Object.get` returns array instead of object if key name is number-like

### LINKED ISSUE #2198: Object attributes transformed in array
### New Issue Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/Parse-SDK-JS/issues?q=is%3Aissue).
- [x] I can reproduce the issue with the latest versions of [Parse Server](https://github.com/parse-community/parse-server/releases) and the [Parse JS SDK](https://github.com/parse-community/Parse-SDK-JS/releases). 

### Issue Description
When I get an Object attribute on my object with Parse 5.2, it returns an array instead of the expected object.
(Parse 5.1 return an Object)

### Steps to reproduce
Get an Object attribute with parse 5.2

Server
- Parse Server version: `7.1`

Database
- System (MongoDB or Postgres): `MongoDB`
- Database version: `7`
- Local or remote host (MongoDB Atlas, mLab, AWS, Azure, Google Cloud, etc): `Local`

Client
- Parse JS SDK version: `5.2`

### Logs

Parse 5.1:

Parse 5.2:

### PR DESCRIPTION
**Pull request by mtrezza**
## Pull Request

- Report security issues [confidentially](https://github.com/parse-community/Parse-SDK-JS/security/policy).
- Any contribution is under this [license](https://github.com/parse-community/Parse-SDK-JS/blob/alpha/LICENSE).
- Link this pull request to an [issue](https://github.com/parse-community/Parse-SDK-JS/issues?q=is%3Aissue).

## Issue

Closes: https://github.com/parse-community/Parse-SDK-JS/issues/2198

## Approach

Issue demo

## Tasks

- [x] Add tests

### CODE REVIEW

File: src/ObjectStateMutations.ts [hunk 1]
```diff
@@ -191,7 +191,7 @@ export function commitServerChanges(
       typeof val === 'object' &&
       !Array.isArray(val) &&
       Object.keys(val).length > 0 &&
-      Object.keys(val).some(k => !isNaN(parseInt(k))) &&
+      Object.keys(val).some(k => k === String(Number(k)) && Number.isInteger(Number(k))) &&
```
Comment (mtrezza) [inline src/ObjectStateMutations.ts]:
```suggestion
      Object.keys(val).every(k => k === String(Number(k)) && Number.isInteger(Number(k))) &&
```

File: integration/test/ParseObjectTest.js [hunk 2]
```diff
@@ -2056,6 +2056,39 @@ describe('Parse Object', () => {
...
+  fit('returns correct field values', async () => {
+    const values = [
+      // { field: 'string', value: 'string' },
+      // { field: 'number', value: 1 },
+      // { field: 'boolean', value: true },
+      // { field: 'array', value: [1, 2, 3] },
+      // { field: 'object', value: { key: 'value' } },
+      // { field: 'object', value: { key1: 'value1', key2: 'value2' } },
+      // { field: 'object', value: { key1: 1, key2: 2 } },
+      { field: 'object', value: { '1x1': 1 } },
```
Comment (mtrezza) [inline integration/test/ParseObjectTest.js:2070]:
```suggestion
      { field: 'object', value: { '1x1': 1 } },
      { field: 'object', value: { '1x1': 1, '2': 2 } },

```

File: integration/test/ParseObjectTest.js [hunk 3]
```diff
@@ -2056,6 +2056,40 @@ describe('Parse Object', () => {
...
+      { field: 'object', value: { '1x1': 1, '2': 2 } },
+      { field: 'object', value: { '1': 1 } },
```
Comment (mtrezza) [inline integration/test/ParseObjectTest.js:2073]:
```suggestion
      { field: 'object', value: { '0': 0 } },
      { field: 'object', value: { '1': 1 } },
```

**Comment by mkmandar123** [comment]
@mtrezza this fix would return wrong output e.g.
1. `{ field: 'object', value: { '1x1': 1, '2': 2, '3': 3 } },` return as jsonArray which is wrong
2. `{ field: 'object', value: { '1': 1 } },` return as jsonArray which is also wrong as `0` index is missing
 I have created a PR https://github.com/parse-community/Parse-SDK-JS/pull/2206/ which considers these cases as well.

**Comment by mtrezza** [comment]
An index of 0 is not required for an array to be an array. See https://github.com/parse-community/Parse-SDK-JS/issues/2198#issuecomment-2210725825.

File: src/ObjectStateMutations.ts [hunk 4]
```diff
@@ -191,7 +191,7 @@ export function commitServerChanges(
       typeof val === 'object' &&
       !Array.isArray(val) &&
       Object.keys(val).length > 0 &&
-      Object.keys(val).some(k => !isNaN(parseInt(k))) &&
+      Object.keys(val).every(k => k === String(Number(k)) && Number.isInteger(Number(k))) &&
```
Comment (dplewis) [inline src/ObjectStateMutations.ts]: Add another check for dot notation arrays.

A dot notation on array looks like `items.0.count` just check for a dot and an index.

The result from the server would return something like

`{ items: { '0': { count: 20 } } }`

**Review by dplewis — CHANGES_REQUESTED** [review]

File: src/ObjectStateMutations.ts (hunk 4 above)
Comment (dplewis) [reply src/ObjectStateMutations.ts:194]: @mtrezza Can you make dot notation on arrays experimental?

`CoreManager.set('ENABLE_DOT_NOTATION_ARRAY', true)`

File: src/__tests__/ObjectStateMutations-test.js [hunk 5]
```diff
@@ -313,16 +313,6 @@ describe('ObjectStateMutations', () => {
...
-  it('can commit nested json array changes from the server to empty serverData', () => {
-    const serverData = {};
-    const objectCache = {};
-    ObjectStateMutations.commitServerChanges(serverData, objectCache, {
-      items: { '0': { count: 20 }, '1': { count: 5 } },
-    });
-    expect(serverData).toEqual({ items: [{ count: 20 }, { count: 5 }] });
-    expect(objectCache).toEqual({ items: '[{"count":20},{"count":5}]' });
-  });
-
```
Comment (dplewis) [inline src/__tests__/ObjectStateMutations-test.js:325]: @mtrezza I would keep this to show that the JSON array doesn't transform

```
it('can commit nested json array changes from the server to empty serverData', () => {
    const serverData = {};
    const objectCache = {};
    ObjectStateMutations.commitServerChanges(serverData, objectCache, {
      items: { '0': { count: 20 }, '1': { count: 5 } },
    });
    // Should not transform
    expect(serverData).toEqual({ items: { '0': { count: 20 }, '1': { count: 5 } } });
    expect(objectCache).toEqual({ items: '{"0":{"count":20},"1":{"count":5}}' });
  });
```

**Review by dplewis — CHANGES_REQUESTED** [review]

File: src/__tests__/ObjectStateMutations-test.js (hunk 5 above)
Comment (mtrezza) [reply src/__tests__/ObjectStateMutations-test.js:325]: This test fails after removing the JSON array check, that's why I've removed it

Comment (dplewis) [reply src/__tests__/ObjectStateMutations-test.js:325]: Look again I updated it, use the test I posted. The ensures the JSON Array doesn't get converted to array

Comment (mtrezza) [reply src/__tests__/ObjectStateMutations-test.js:325]: Added the new test

**Comment by mtrezza** [comment]
@dplewis anything else that should be removed from the changes in https://github.com/parse-community/Parse-SDK-JS/pull/2120/files? Or can we merge

**Comment by dplewis** [comment]
Thats it, this is ready to merge

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.3.0-alpha.2](https://github.com/parse-community/Parse-SDK-JS/releases/tag/5.3.0-alpha.2)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.3.0-beta.1](https://github.com/parse-community/Parse-SDK-JS/releases/tag/5.3.0-beta.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.3.0](https://github.com/parse-community/Parse-SDK-JS/releases/tag/5.3.0)

**Pull request merged**